lib/filesystem.nix

```diff
@@ -148,10 +159,131 @@ in
     # The path to recursively list
     dir:
     lib.flatten (lib.mapAttrsToList (name: type:
-    if type == "directory" then
-      lib.filesystem.listFilesRecursive (dir + "/${name}")
-    else
-      dir + "/${name}"
+      if type == "directory" then
+        lib.filesystem.listFilesRecursive (dir + "/${name}")
+      else
+        dir + "/${name}"
   ) (builtins.readDir dir));
 
+  /*
+    Transform a directory tree containing package files suitable for
+    `callPackage` into a matching nested attribute set of derivations.
+
+    For a directory tree like this:
+
+    ```
+    my-packages
+    ├── a.nix
+    ├── b.nix
+    ├── c
+    │  ├── my-extra-feature.patch
+    │  ├── package.nix
+    │  └── support-definitions.nix
+    └── my-namespace
+       ├── d.nix
+       ├── e.nix
+       └── f
+          └── package.nix
+    ```
+
+    `packagesFromDirectory` will produce an attribute set like this:
+
+    ```nix
+    # packagesFromDirectory {
+    #   callPackage = pkgs.callPackage;
+    #   directory = ./my-packages;
+    # }
+    {
+      a = pkgs.callPackage ./my-packages/a.nix { };
+      b = pkgs.callPackage ./my-packages/b.nix { };
+      c = pkgs.callPackage ./my-packages/c/package.nix { };
+      my-namespace = {
+        d = pkgs.callPackage ./my-packages/my-namespace/d.nix { };
+        e = pkgs.callPackage ./my-packages/my-namespace/e.nix { };
+        f = pkgs.callPackage ./my-packages/my-namespace/f/package.nix { };
+      };
+    }
+    ```
+
+    In particular:
+    - `.nix` files are returned with the `.nix` extension removed.
+    - Other files are ignored.
+    - If a directory contains a `package.nix`, the directory name is used as
+      the package name.
+    - Otherwise, directories are recursively traversed.
+
```

Comment:
Is this a nixpkgs restriction elsewhere? I think it would require `package/package.nix` with this code, but it should be possible.

Reply:
Ah yeah that works, though it can easily trip up people when they start out with
```
nested
├── a.nix
└── b.nix
```

But then add a `package.nix`, and suddenly the output structure changes, ignoring the other files.

However, if it's not a nested package set, `package.nix` does actually work just fine to declare `package` in the current implementation! It might make sense to flip around the recursion to make this consistent, such that if the `directory` itself contains a `package.nix`, it turns into a single package, not a set.

Reply:
> It might make sense to flip around the recursion to make this consistent, such that if the `directory` itself contains a `package.nix`, it turns into a single package, not a set.

I think this would have to be a special case because then `packagesFromDirectoryRecursive` wouldn't return an attribute set.

Reply:
Actually I managed to make it work rather nicely. PTAL